collated/main/A0147967J.md

```diff
@@ -0,0 +1,1224 @@
+
+    }
+
+}
+```
+###### \java\seedu\address\logic\parser\Parser.java
+``` java
+    private Command prepareBlock(String args) {
+    	Matcher matcher = BLOCK_DATA_ARGS_FORMAT.matcher(args.trim());
+        if (!matcher.matches()) {
+            return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, BlockCommand.MESSAGE_USAGE));
+        }
+        try {
+            
+            String startInput = matcher.group("startTime");
+            String endInput = matcher.group("endTime");
+            
+            return new BlockCommand(
+                    getTagsFromArgs(matcher.group("tagArguments")),
+                    new TaskDate(getDateFromString(startInput).getTime()),
+                    new TaskDate(getDateFromString(endInput).getTime())
+                    );
+        } catch (IllegalValueException ive) {
+            return new IncorrectCommand(ive.getMessage());
+        }
+	}
+```
+###### \java\seedu\address\logic\parser\Parser.java
+``` java
+    /**
+     * Parses arguments in the context of the complete task command.
+     *
+     * @param args full command args string
+     * @return the prepared command
+     */
+    private Command prepareComplete(String args) {
+
+        Optional<Integer> index = parseIndex(args);
+        if(!index.isPresent()){
+            return new IncorrectCommand(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, CompleteCommand.MESSAGE_USAGE));
+        }
+
+        return new CompleteCommand(index.get());
+    }
+```
+###### \java\seedu\address\logic\parser\Parser.java
+``` java
+    /**
+     * Returns the view command with input date parsed 
+     * @param arguments passed by user
+     * @return prepared view command
+     */
+    private Command prepareView(String arguments) {
+		// TODO Auto-generated method stub
+    	Date date;
+    	try{
+    		date = getDateFromString(arguments);
+    	}catch(IllegalArgumentException e){
+    		return new IncorrectCommand(e.getMessage());
+    	}
+		return new ViewCommand(new TaskDate(date));
+	}
+```
+###### \java\seedu\address\logic\URManager.java
+``` java
+/**
+ * Stores and provides context for undo/redo operations.
+ */
+public class URManager {
+	
+	private ArrayDeque<Context> undoQueue;
+	private ArrayDeque<Context> redoQueue;
+	
+	/** Arbitrarily chosen number to ensure overall performance. */
+	private final int MAX_TIMES = 3;
+	
+	public URManager(){		
+		undoQueue = new ArrayDeque<Context>();
+		redoQueue = new ArrayDeque<Context>();		
+	}
+	
+	/**
+	 * Adds the command to undo queue for LogicManager.
+	 */
+	public void addToUndoQueue(Model model, Command command){
+		if(!isUndoable(command)){
+			//Stop here to wait for result.
+		}else{
+			if(!isIgnored(command)){
+				if(undoQueue.size() == MAX_TIMES) undoQueue.removeFirst();
+				undoQueue.addLast(new Context(model, command));
+				redoQueue.clear();
```

Comment:
Minimize arrowhead anti pattern https://lostechies.com/chrismissal/2009/05/27/anti-patterns-and-worst-practices-the-arrowhead-anti-pattern/